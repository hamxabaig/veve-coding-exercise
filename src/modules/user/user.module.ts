import { Logger, Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { User } from 'modules/user/models/user.model';
import { UserService } from 'modules/user/services/user.service';

@Module({
  imports: [TypeOrmModule.forFeature([User])],
  exports: [TypeOrmModule, UserService],
  providers: [Logger, UserService],
})
export class UserModule {}
