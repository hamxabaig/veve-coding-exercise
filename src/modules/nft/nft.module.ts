import { Logger, Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { NFT } from 'modules/nft/models/nft.model';
import { NFTTransfer } from 'modules/nft/models/nft_transfer.model';
import { UserModule } from 'modules/user/user.module';
import { NftResolver } from './graphql/nft.resolver';
import { NftService } from './services/nft.service';

@Module({
  imports: [TypeOrmModule.forFeature([NFT, NFTTransfer]), UserModule],
  providers: [Logger, NftResolver, NftService],
})
export class NftModule {}
