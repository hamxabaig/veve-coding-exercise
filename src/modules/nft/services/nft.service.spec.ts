import { Logger } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { NFT } from 'modules/nft/models/nft.model';
import { NFTTransfer } from 'modules/nft/models/nft_transfer.model';
import { UserService } from 'modules/user/services/user.service';
import { NftService } from './nft.service';

export const nftRepositoryMock = {
  findOneBy: jest.fn(),
  save: jest.fn(),
};
export const nftTransferRepositoryMock = {
  create: jest.fn(),
  save: jest.fn(),
};
export const userServiceMock = {
  findById: jest.fn(),
};
export const loggerMock = {
  log: jest.fn(),
};

describe('NFTService', () => {
  let service: NftService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        NftService,
        { provide: getRepositoryToken(NFT), useValue: nftRepositoryMock },
        {
          provide: getRepositoryToken(NFTTransfer),
          useValue: nftTransferRepositoryMock,
        },
        {
          provide: UserService,
          useValue: userServiceMock,
        },
        {
          provide: Logger,
          useValue: loggerMock,
        },
      ],
    }).compile();

    service = module.get<NftService>(NftService);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('transfers nft to new owner', async () => {
    userServiceMock.findById.mockResolvedValue({ id: 2 });
    nftRepositoryMock.findOneBy.mockResolvedValue({ owner: { id: 1 } });
    nftTransferRepositoryMock.save.mockResolvedValue({ id: 1 });

    const transferId = await service.transfer(1, 1, 2);
    expect(transferId).toBe(1);
  });

  it('throws error if nft does not belong to owner', async () => {
    userServiceMock.findById.mockResolvedValue({ id: 2 });
    nftRepositoryMock.findOneBy.mockResolvedValue({ owner: { id: 1 } });

    const transferFn = () => service.transfer(1, 2, 2);
    expect(transferFn).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error if nft was not found', async () => {
    userServiceMock.findById.mockResolvedValue({ id: 2 });
    nftRepositoryMock.findOneBy.mockResolvedValue(undefined);

    const transferFn = () => service.transfer(1, 2, 2);
    expect(transferFn).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error if new owner was not found', async () => {
    userServiceMock.findById.mockResolvedValue(undefined);
    nftRepositoryMock.findOneBy.mockResolvedValue({ owner: { id: 1 } });

    const transferFn = () => service.transfer(1, 2, 2);
    expect(transferFn).rejects.toThrowErrorMatchingSnapshot();
  });
});
