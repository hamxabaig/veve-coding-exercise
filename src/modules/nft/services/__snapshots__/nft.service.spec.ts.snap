// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTService throws error if new owner was not found 1`] = `"User not found with id 2"`;

exports[`NFTService throws error if nft does not belong to owner 1`] = `"NFT not found with id 1 or you might not have access to it"`;

exports[`NFTService throws error if nft was not found 1`] = `"NFT not found with id 1 or you might not have access to it"`;
